File: Iteration1-Java/src/main/java/cse3063group3/Student.java
package cse3063group3;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import org.json.*;
import java.util.Arrays;

public class Student extends Person {

    final int MAX_COURSES = 5; // add this to the uml
    private HashMap<CourseSection, String> courseSectionsOfStudent; // this will be map
    //private HashMap<Course, Integer> selectedSections = new HashMap<>();    // TODO: Depriciated delete later.
    private Advisor advisor;


    /**
     * 
     * @param id                 ID of the student.
     * @param name               Name of the student.
     * @param eMail              E-Mail address of the student.
     * @param password           Password of the student.
     * @param controllerInstance Referance to the controller instance.
     */
    public Student(int id, String name, String eMail, String password) {
        super(id, name, eMail, password);
        this.courseSectionsOfStudent = new HashMap<CourseSection, String>();
        this.advisor = null;
    }

    // add the course and set it's status to selected
    // called by the
    public void addCourseSection(CourseSection courseSec) {
        int previouslyChosenCourseCount = this.getStudentCoursesWithStatus("pending").size()
                + this.getStudentCoursesWithStatus("selected").size()
                + this.getStudentCoursesWithStatus("approved").size()
                + this.getStudentCoursesWithStatus("waiting").size();

        if (previouslyChosenCourseCount < MAX_COURSES) {
            if (this.CheckScheduleAvailable(courseSec)){
                this.courseSectionsOfStudent.put(courseSec, "selected");
                // TODO: Also add student to coursesections students list.
            }
        }
    }

    public void changeStatus(CourseSection courseSection, String status) {
        if (status.equals("selected"))
            addCourseSection(courseSection);
        this.courseSectionsOfStudent.replace(courseSection, status);
    }


    /**
     * <b>Probably Depriciated IDK!</b> <p>
     * Finds all courses from the current student instance that match the given
     * argument status and returns it as an ArrayList.
     * 
     * @param status String to match course status against.
     * @return A list of courses that match with the given status string.
     *         <p>
     *         If "status" is null or left blank, returns all courses.
     */
    private ArrayList<Course> getStudentCoursesWithStatus(String status) {
        ArrayList<CourseSection> courseSectionsList = new ArrayList<>();
        ArrayList<Course> coursesList = new ArrayList<>();

        if (status == null || status.isBlank()) {
            courseSectionsList.addAll(this.courseSectionsOfStudent.keySet());
            courseSectionsList.forEach((courseSection) -> {coursesList.add(courseSection.getCourse());});
            return coursesList;
        }

        for (Map.Entry<CourseSection, String> entry : this.courseSectionsOfStudent.entrySet()) {
            if (entry.getValue().equals(status)){
                coursesList.add(entry.getKey().getCourse());
            }
        }
        return coursesList;
    }

    /**
     * Finds all course sections from the current student instance that match the given
     * argument status and returns it as an ArrayList.
     * 
     * @param status String to match course section status against.
     * @return A list of course sections that match with the given status string.
     *         <p>
     *         If "status" is null or left blank, returns all course sections.
     */
    private ArrayList<CourseSection> getStudentCourseSectionsWithStatus(String status) {
        ArrayList<CourseSection> courseSectionsList = new ArrayList<>();

        if (status == null || status.isBlank()) {
            courseSectionsList.addAll(this.courseSectionsOfStudent.keySet());
            return courseSectionsList;
        }

        for (Map.Entry<CourseSection, String> entry : this.courseSectionsOfStudent.entrySet()) {
            if (entry.getValue().equals(status)){
                courseSectionsList.add(entry.getKey());
            }
        }
        return courseSectionsList;
    }

    /**
     * <pre>
     * Tries to print a numbered list of courses taken by the student that match given status.
     * </pre>
     * 
     * @param status String to match course status against.
     *               <p>
     *               If left empty prints all courses taken by the student without
     *               matching any status.
     */
    public void listStudentCoursesWithStatus(String status) {
        ArrayList<String> prefixText = new ArrayList<>();
        if (status == null) {
            status = "";
        }

        ArrayList<CourseSection> courseSectionsWithStatusList = getStudentCourseSectionsWithStatus(status);

        if (status.isBlank()) {
            prefixText.add("All courses taken by " + this.getName() + ".\n");
        } else {
            // Ex: Barış Sedefoğlu's finished courses.
            prefixText.add(this.getName() + "'s " + status + " courses.\n");
        }

        if (courseSectionsWithStatusList == null || courseSectionsWithStatusList.isEmpty()) {
            // Ex: Student: Barış Sedefoğlu with ID: 150120002 currently doesn't have any
            // courses.
            prefixText.add("Student: " + this.getName() + " with ID: " + this.getId()
                    + " currently doesn't have any " + status + " courses.\n");
        } else {
            for (int i = 0; i < courseSectionsWithStatusList.size(); i++) {
                // Ex: 3- Object Oriented Software Design, Status: taken, Credits: 5, Course ID: 105
                prefixText.add((i + 1) + "- " + courseSectionsWithStatusList.get(i).getCourse().getName() + "." + courseSectionsWithStatusList.get(i).getCourseSectionNumber() 
                                + ", Status: " + this.courseSectionsOfStudent.get(courseSectionsWithStatusList.get(i))
                                + ", Credits: " + courseSectionsWithStatusList.get(i).getCourse().getCredits()
                                + ", Course ID: " + courseSectionsWithStatusList.get(i).getCourse().getId()
                                + "\n");
            }
        }
        this.getControllerInstance().menuOperation(prefixText);
    }

    @Override
    public void mainMenu() {
        ArrayList<String> prefixText = new ArrayList<>(Arrays.asList("\n\nStudent system.\n",
                "Available options:\n"));

        ArrayList<String> optionsText = new ArrayList<>(Arrays.asList("List offered courses in this semester.",
                "List pending courses.",
                "List approved courses.",
                "List finished courses.",
                "List selected courses.",
                "Select Courses.",
                "Send selected courses for advisor approval.",
                "Show weekly schedule",
                "Exit.\n"));

        ArrayList<String> suffixText = new ArrayList<>(Arrays.asList("Please choose an option: "));
        // Loop inside the student menu until exit is chosen.
        while (true) {
            // Send the main menu body to be printed and get an option from user.
            int choice = this.getControllerInstance().menuOperation(prefixText, optionsText, suffixText);

            if (choice == optionsText.size() - 1) {
                this.getControllerInstance()
                .menuOperation(new ArrayList<>(Arrays.asList("You are leaving the system...\n")));
                break;
                // student exit the system
            }

            switch (choice) {
                case 0 -> // "List offered courses in this semester."
                    // Prints with course status information for current student.
                    this.getControllerInstance().printAllDepartmentCourses(this.courseSectionsOfStudent);
                case 1 -> // "List pending courses."
                    listStudentCoursesWithStatus("pending");
                case 2 -> // "List approved courses."
                    listStudentCoursesWithStatus("approved");
                case 3 -> // "List finished courses."
                    listStudentCoursesWithStatus("finished");
                case 4 -> // "List selected courses."
                    listStudentCoursesWithStatus("selected");
                case 5 -> // "Select Courses."
                    selectCourseSectionsForSemester();
                case 6 -> {
                    // "Send selected courses to advisor approval."
                    this.getControllerInstance()
                            .menuOperation(new ArrayList<>(Arrays.asList("All selected courses are sended to advisor approval.\n")));
                    this.courseSectionsOfStudent.replaceAll((course, status) -> status.equals("selected") ? "pending" : status);
                }
                case 7 -> this.printWeeklySchedule();
                default ->//TODO: burada try-catch kullanılmalı, mesela faladan enter basarsa kişi bu mesajı görmemeli
                    this.getControllerInstance().menuOperation(new ArrayList<>(Arrays.asList("Error: This shouldn't happen. StudentSystem.studentSystem()")));
            }
        }
    }

    /**
     * <pre>
     * Numbers and prints all of the available courses in the department for course selection.
     * Upon selection(s) does course prerequisite checks and tries to add each selected course to student's courses HashMap.
     * </pre>
     */
    private void selectCourseSectionsForSemester() {
        int previouslyChosenCourseCount = this.getStudentCoursesWithStatus("pending").size()
                + this.getStudentCoursesWithStatus("selected").size()
                + this.getStudentCoursesWithStatus("approved").size()
                + this.getStudentCoursesWithStatus("waiting").size();
                
        while (true) {
            int sectionSelection = 1;
            Course chosenCourse = this.getControllerInstance().selectCourseFromAllDepartmentCourses(/*this.getName(),*/ this.getAllCourses());
            //Chose exit.
            if (chosenCourse == null) {
                break;
            }

            if (this.MAX_COURSES <= previouslyChosenCourseCount + 1) {
                this.getControllerInstance().menuOperation(new ArrayList<>(Arrays.asList("Course is not selected because at most " + MAX_COURSES + " courses can be selected in a semester.")));
                break;
            }

            if (getAllCourses().containsKey(chosenCourse)) {
                // TODO: statüye göre mesaj yazdırılabilir
                this.getControllerInstance().menuOperation(new ArrayList<>(Arrays.asList(chosenCourse.getName() + " is already in your courses. Please choose another course.\n")));
            } else {
                // check if prerequisites are met
                if (chosenCourse.CheckPrerequisitesMet(this.getStudentCoursesWithStatus("finished"), this)) {
                    if (chosenCourse.getCourseSections().size() > 1) {
                        this.getControllerInstance().menuOperation(new ArrayList<>(Arrays.asList("Please select a section.")));
                        for (int j = 0; j < chosenCourse.getCourseSections().size() - 1; j++) {
                            this.getControllerInstance().menuOperation(new ArrayList<>(Arrays.asList((j + 1) + "- " + chosenCourse.getName() + "." + (j + 1))));
                        }
                        sectionSelection = this.getControllerInstance().getNextInt();
                    }

                    courseSectionsOfStudent.put(chosenCourse.getCourseSections().get(sectionSelection - 1), "selected");
                    previouslyChosenCourseCount++;
                } else {
                    // TODO: buraya gerek yok. Course classının içinde yazdırıyoruz zaten
                    // System.out
                    // .println(chosenCourse.getName() + " is not selected beacuse not all
                    // prereguisites are met");
                }
            }
        }
    }

    public void setCourseSectionStatusToApprovedOrRejected() {
        ArrayList<CourseSection> pendingCourseSections;

        while (true) {
            pendingCourseSections = this.getStudentCourseSectionsWithStatus("pending");
            this.getControllerInstance().menuOperation(new ArrayList<>(Arrays.asList("\n\n")));
            this.listStudentCoursesWithStatus("pending");  // This isn't really meant as an input method so it doesn't print the exit option.
            this.getControllerInstance().menuOperation(new ArrayList<>(Arrays.asList((pendingCourseSections.size() + 1) + "- Exit.\n\n", "Please select a course section: ")));
            
            int result = this.getControllerInstance().getNextInt() - 1;
            if (result >= 0 && result < pendingCourseSections.size()) {
                ArrayList<String> prefixText = new ArrayList<>(
                        Arrays.asList("\n\nApprove - Reject " + pendingCourseSections.get(result).getName() + "\n"));

                ArrayList<String> optionsText = new ArrayList<>(Arrays.asList("Approve.",
                        "Reject",
                        "Skip\n"));

                ArrayList<String> suffixText = new ArrayList<>(Arrays.asList("Please choose an option: "));
                while (true) {
                    int option = this.getControllerInstance().menuOperation(prefixText, optionsText, suffixText);

                    if (option == 0) {
                        this.changeStatus(pendingCourseSections.get(result), "approved");
                        // TODO: Add student object to course / coursesection.
                        break;
                    }
                    if (option == 1) {
                        this.changeStatus(pendingCourseSections.get(result), "rejected");
                        break;
                    } 
                    if (option == 2) {
                        break;
                    }

                    //Exhausted all possibilities. The choice is invalid.
                    this.getControllerInstance().menuOperation(new ArrayList<>(Arrays.asList("Please choose a valid option.\n")));

                }
            } else {
                if (result < 0)
                    this.getControllerInstance().menuOperation(new ArrayList<>(Arrays.asList("Please choose a valid option.\n")));
                if (result == pendingCourseSections.size()) {
                    break;
                }
            }
        }
    }

    /**
     * 
     * @param courseSection chekcs against this
     * @return
     */
    public boolean CheckScheduleAvailable(CourseSection courseSectionToCheckAgainst) {

        ArrayList<CourseSection> takenCourseSectionList = new ArrayList<>();
        takenCourseSectionList.addAll(getStudentCourseSectionsWithStatus("selected"));
        takenCourseSectionList.addAll(getStudentCourseSectionsWithStatus("pending"));
        takenCourseSectionList.addAll(getStudentCourseSectionsWithStatus("approved"));
        takenCourseSectionList.addAll(getStudentCourseSectionsWithStatus("waiting"));
        // TODO: course status will change to waiting after selection when course is full.
        ArrayList<Integer> schedule = new ArrayList<Integer>();

        for (CourseSection courseSection : takenCourseSectionList) {
            ArrayList<LectureHour> lectureHours = courseSection.getLectureHours();
            for (LectureHour lectureHour : lectureHours) {
                schedule.add(lectureHour.getLectureTimeTableNumber());
            }
        }
        
        ArrayList<Integer> newCourseHours = new ArrayList<Integer>();
        for(LectureHour lectureHour : courseSectionToCheckAgainst.getLectureHours()){
            newCourseHours.add(lectureHour.getLectureTimeTableNumber());
        }

        if(schedule.stream().anyMatch(newCourseHours::contains)){
            return false;
        }


        return true;
    }
    
    public void printWeeklySchedule() {

        ArrayList<CourseSection> takenCourseSectionList = new ArrayList<>();
        takenCourseSectionList.addAll(getStudentCourseSectionsWithStatus("selected"));
        takenCourseSectionList.addAll(getStudentCourseSectionsWithStatus("pending"));
        takenCourseSectionList.addAll(getStudentCourseSectionsWithStatus("approved"));
        takenCourseSectionList.addAll(getStudentCourseSectionsWithStatus("waiting"));
        // TODO: course status will change to waiting after selection when course is full.
        String[][] schedule = new String[7][10];
        int size = takenCourseSectionList.size();
        for (int i = 0; i < size; i++) {
            ArrayList<LectureHour> lectureHours = takenCourseSectionList.get(i).getLectureHours();
            for (int j = 0; j < lectureHours.size(); j++) {
                // first dimansion of array takes lecture day as index, second takes lecture
                // hour
                schedule[lectureHours.get(j).getLectureDay()][lectureHours.get(j).getLectureHourNumber()] = 
                    /*takenCourse.get(j).getName();*/ takenCourseSectionList.get(i).getCourse().getId() + "." + takenCourseSectionList.get(i).getCourseSectionNumber() + "";
                // TODO: ^ I highly doubt you can fit the course name in the schedule.
                // 'Data Structures and Algorithms' will easily mess up the entire formatting.
                // Changing to getID. If getName was intentional. Just uncomment. Barış
            }
        }

        ArrayList<String> scheduleTextList = new ArrayList<>();
        for (int j = 0; j < 10; j++) {
            StringBuilder stringBuilder = new StringBuilder();
            for (int i = 0; i < 7; i++) {
                if (schedule[i][j] == null) {
                    stringBuilder.append("     -     ");
                    // Adding continue here otherwise later on schedule[i][j] will print "null" in the schedule.
                    // I don't think this is intentional. If it was, just delete this continue. Barış
                    continue;
                }
                stringBuilder.append(" " + schedule[i][j] + " ");   // The last spacer was 3 whitespaces, changed to 1 otherwise it wont match the above spacer.
                                                                    // "         " has nine; courseId has 7 so it can be padded 1 space on each side.
                                                                    // Undo if it was intentional. Barış.
            }
            stringBuilder.append("\n");
            scheduleTextList.add(stringBuilder.toString());
        }
        this.getControllerInstance().menuOperation(scheduleTextList);
    }






    /**
     * <b>Probably Depriciated IDK!</b> <p>
     * @return Returns a HashMap<Course, String> containing the Course - Course Status relations.
     */
    public HashMap<Course, String> getAllCourses() {
        var coursesMap = new HashMap<Course, String>();
        this.courseSectionsOfStudent.forEach((a, b) -> {coursesMap.put(a.getCourse(), b); });

        return coursesMap;
    }

    /**
     * @return Returns the Student.courseSectionsofStudent field which contains the CourseSection - CourseSection Status relations.
     */
    public HashMap<CourseSection, String> getAllCourseSections() {
        return this.courseSectionsOfStudent;
    }

    public Advisor getAdvisor() {
        return (this.advisor);
    }

    @Override
    public int getId() {
        return (super.getId());
    }

    @Override
    public String getName() {
        return (super.getName());
    }

    public void setAdvisor(Advisor advisor) {
        this.advisor = advisor;
    }

    public JSONObject toJSON() {
        return null;
    };

    public void setCourseSectionStatus(CourseSection courseSection, String status) {
        this.courseSectionsOfStudent.put(courseSection, status);
    }

    public boolean hasPendingCourse() {
        return this.courseSectionsOfStudent.containsValue("pending");
    }
}

File: Iteration2-Python/src/Student.py

from typing import List, Dict
import math
from Person import Person
from Advisor import Advisor
from LectureHour import LectureHour
from CourseSection import CourseSection

import logging
logger = logging.getLogger(__name__)    # to ensure that the logger is named after the module
class Student(Person):
    MAX_COURSES = 5

    def __init__(self, id : int, name : str, e_mail : str, password : str, semester : int, controller_instance):
        super(Student, self).__init__(id, name, e_mail,password, controller_instance)
        logger.info(f"Super type' 'Person' constructor is called. {__name__} type object is created.")
        # set parameters initialized inside of the super type constructor as private
        self.__id : int = id
        self.__name : str = name
        self.__e_mail : str = e_mail
        self.__password : str = password
        self.__controller_instance = controller_instance

        self.__advisor : Advisor = None
        self.__semester : int = semester
        self.__course_sections_of_student : Dict = {}
        self.__course_sections_of_student_with_letter_grades : Dict = {}    # course section(CourseSection) - letter grade(str)

        from Transcript import Transcript
        self.__transcript : Transcript = Transcript(semester, self.__course_sections_of_student_with_letter_grades)
    
    
    def get_student_course_sections_with_status(self, status : str = None) -> List:
        if (self.__course_sections_of_student is None):
            logger.warning(f"Student with the id of {self.get_id()} does not have any course section.")
            return None
        
        if(status is None ):
            course_section_list = []
            for course_section in self.__course_sections_of_student.keys():
                course_section_list.append(course_section)
            return course_section_list
        else:
            course_section_list = []
            for course_section , stat in self.__course_sections_of_student.items():
                if status == stat:
                    course_section_list.append(course_section)
            return course_section_list


    def add_course_section(self ,course_section) -> None:
        previously_chosen_course_count = len(self.get_student_course_sections_with_status('pending')) 
        + len(self.get_student_course_sections_with_status('approved'))
        + len(self.get_student_course_sections_with_status('waiting'))
        + len(self.get_student_course_sections_with_status('selected'))
        
        logger.debug(f"Checking previously chosen course count, result = {previously_chosen_course_count}")
        if previously_chosen_course_count <= self.MAX_COURSES:
            if self.check_schedule_available(course_section):
                self.__course_sections_of_student[course_section] = 'selected'


    def change_status(self, course_section, status) -> None:
        if(status == 'selected'):
            logger.info(f"Course section{course_section.get_id()} is added to student' {self.get_id()} course sections.")
            self.add_course_section(course_section)
        logger.info(f"Updating course' status to {status}.")
        self.__course_sections_of_student[course_section] = status


    def list_student_courses_with_status(self, status = None) -> None:
        prefix_text : List = []

        course_section_with_status_list = self.get_student_course_sections_with_status(status)
        if status is None:
            prefix_text.append("All courses taken by " + self.get_name() + ".\n")
        else:
            prefix_text.append(self.get_name() + "'s " + status + " courses.\n")
        
        if(len(course_section_with_status_list) == 0):
            prefix_text.append("Student: " + self.get_name() + " with ID: " + str(self.get_id())
                    + " currently doesn't have any " + status + " courses.\n")
        else:
            i = 1
            for course_section in course_section_with_status_list:
                prefix_text.append(f"{i:^3}| "+ f"{course_section.get_course().get_name():<40}| " + f"{course_section.get_id()}"
                                + ", Status: " + self.__course_sections_of_student[course_section]
                                + ", Credits: " + str(course_section.get_course().get_credits())
                                + ", Course ID: " + str(course_section.get_course().get_id())
                                + "\n")
                i += 1
        self.get_controller_instance().menu_operation(prefix_text)


    def main_menu(self) -> None:
        prefix_text = ["\n\nStudent system.\n",
                        "Available options:\n\n"]
        options_text = ["List offered courses in this semester.",
                "List pending courses.",
                "List approved courses.",
                "List finished courses.",
                "List selected courses.",
                "Select Courses.",
                "Send selected courses for advisor approval.",
                "Show weekly schedule",
                "Show transcript information.",
                "Exit.\n"]
        suffix_text = ["Please choose an option: \n"]

        while(True):
            choice = self.get_controller_instance().menu_operation(prefix_text, options_text ,suffix_text)

            if choice == len(options_text) -1:
                self.get_controller_instance().menu_operation(["You are leaving the system...\n"])
                break
            match choice:
                case 0:
                    self.get_controller_instance().print_all_department_courses(self.__course_sections_of_student)
                case 1:
                    self.list_student_courses_with_status("pending")
                case 2:
                    self.list_student_courses_with_status("approved")
                case 3:
                    self.list_student_courses_with_status("finished")
                case 4:
                    self.list_student_courses_with_status("selected")
                case 5:
                    self.__select_course_sections_for_semester()
                case 6:
                    self.get_controller_instance().menu_operation(["All selected courses are sended to advisor approval.\n"])
                    for course_section, status in self.__course_sections_of_student.items():
                        if status == 'selected':
                            self.__course_sections_of_student[course_section] = 'pending'
                case 7:
                    self.print_weekly_schedule()
                case 8:
                    self.print_transcript()
                case _:
                    logger.info(f"User entered an option:{choice} which is out of range 1 - {len(options_text)}.")
                    self.get_controller_instance().menu_operation(f"Please enter an integer between 1 and {len(options_text)}\n")
            logger.info(f"Student selected {choice + 1} from the main menu.")


    def __select_course_sections_for_semester(self) -> None:
        previously_chosen_course_count  = len(self.get_student_course_sections_with_status("pending"))
        previously_chosen_course_count += len(self.get_student_course_sections_with_status("selected"))
        previously_chosen_course_count += len(self.get_student_course_sections_with_status("approved"))
        previously_chosen_course_count += len(self.get_student_course_sections_with_status("waiting"))

        while(True):
            section_selection = 1
            chosen_course = self.get_controller_instance().select_a_course_from_all_department_courses(self.get_all_courses())

            if chosen_course is None:
                break

            if self.MAX_COURSES <= previously_chosen_course_count :
                self.get_controller_instance().menu_operation(["Course is not selected because at most " + str(self.MAX_COURSES) + " courses can be selected in a semester."])
                break
            if chosen_course in self.get_all_courses():
                logger.info(f"User tried to take the same course {chosen_course.get_id()} again but system did not allowed it.")
                self.get_controller_instance().menu_operation([chosen_course.get_name() + " is already in your courses. Please choose another course.\n"])
            else:
                if chosen_course.check_prerequisites_met(self):
                    if len(chosen_course.get_course_sections()) > 1:
                        self.get_controller_instance().menu_operation(["Please select a section.\n"])
                        for  i in range(len(chosen_course.get_course_sections())):
                            self.get_controller_instance().menu_operation([str(i + 1) + "- " + chosen_course.get_name() + "."+str(i + 1)+"\n"])
                        
                        while True:
                            try:
                                section_selection = self.get_controller_instance().get_next_int()
                                break
                            except ValueError as e:
                                logging.warning(f"Selection has to be an integer.\n")
                                self.get_controller_instance().menu_operation(f"Selection has to be an integer.\n")
                        
                        
                        self.__course_sections_of_student[chosen_course.get_course_sections()[section_selection - 1]] = 'selected'
                        self.get_controller_instance().menu_operation(f"{chosen_course.get_id()}.{section_selection} is selected successfully.\n")
                        previously_chosen_course_count += 1
                    elif len(chosen_course.get_course_sections()) == 1:
                        self.__course_sections_of_student[chosen_course.get_course_sections()[0]] = 'selected'
                        self.get_controller_instance().menu_operation(f"{chosen_course.get_id()}.{1} is selected successfully.\n")
                        previously_chosen_course_count += 1
                    else:
                        pass
                else: 
                    pass


    def set_course_section_status_to_approved_or_rejected(self) -> None:
        while True:
            pending_course_sections = self.get_student_course_sections_with_status("pending")
            self.get_controller_instance().menu_operation(["\n\n"])
            self.list_student_courses_with_status("pending") # This isn't really meant as an input method so it doesn't print the exit option.
            self.get_controller_instance().menu_operation([str(len(pending_course_sections) + 1) + "- Exit.\n\n", "Please select a course section: "])

            
            while True:
                try:
                    result = self.get_controller_instance().get_next_int() - 1 
                    break
                except ValueError as e:
                    logging.warning(f"Selection has to be an integer.\n")
                    self.get_controller_instance().menu_operation(f"Selection has to be an integer.\n")
            
            if result >= 0 and result <  len(pending_course_sections):
                prefix_text = ["\n\nApprove - Reject " + pending_course_sections[result].get_name() + "\n"]
                options_text = ["Approve.",
                            "Reject",
                            "Skip\n"]
                suffix_text = ["Please choose an option: \n"]

                while True:
                    option = self.get_controller_instance().menu_operation(prefix_text,options_text,suffix_text)

                    if option == 0:
                        self.change_status(pending_course_sections[result], "approved")
                        pending_course_sections[result].add_student(self)
                        break
                    if option == 1:
                        self.change_status(pending_course_sections[result], "rejected")
                        break
                    if option == 2:
                        break

                    # Exhausted all possibilities. The choice is invalid.
                    self.get_controller_instance().menu_operation(["Please choose a valid option.\n"])
            else:
                if result < 0:
                    self.get_controller_instance().menu_operation(["Please choose a valid option.\n"])
                if result == len(pending_course_sections):
                    break


    def check_schedule_available(self, course_section_to_check_against : CourseSection) -> bool:
        taken_course_section_list = []

        taken_course_section_list.extend(self.get_student_course_sections_with_status("selected"))
        taken_course_section_list.extend(self.get_student_course_sections_with_status("pending"))
        taken_course_section_list.extend(self.get_student_course_sections_with_status("approved"))
        taken_course_section_list.extend(self.get_student_course_sections_with_status("waiting"))

        schedule = []

        for course_section in taken_course_section_list:
            lecture_hours = course_section.get_lecture_hours()
            for lecture_hour in lecture_hours:
                schedule.append(lecture_hour.get_lecture_time_table_number())

        new_course_hours  = []
        for lecture_hour in course_section_to_check_against.get_lecture_hours():
            new_course_hours.append(lecture_hour.get_lecture_time_table_number())
        
        for x in schedule:
            for y in new_course_hours:
                if x == y:
                    return False

        return True


    def print_weekly_schedule(self) -> None:
        taken_course_section_list = []
        taken_course_section_list.extend(self.get_student_course_sections_with_status("selected"))
        taken_course_section_list.extend(self.get_student_course_sections_with_status("pending"))
        taken_course_section_list.extend(self.get_student_course_sections_with_status("approved"))
        taken_course_section_list.extend(self.get_student_course_sections_with_status("waiting"))

        schedule = [[None] * 10 for i in range(7)]
        size = len(taken_course_section_list)

        for taken_course_section in taken_course_section_list:
            lecture_hours = taken_course_section.get_lecture_hours()
            for lecture_hour in lecture_hours:
                schedule[lecture_hour.get_lecture_day()][lecture_hour.get_lecture_hour_number()]= str(taken_course_section.get_course().get_id()) + '.'+ str(taken_course_section.get_course_section_number()) 

        schedule_text_list = []

        for i in range(10):
            string_builder = []
            for j in range(7):
                if schedule[j][i] is None:
                    string_builder.append("     -     ")
                    continue
                else:
                    string_builder.append(f"{schedule[j][i]:^11}")
            string_builder.append("\n")
            schedule_text_list.append(' '.join(string_builder))
        
        self.get_controller_instance().menu_operation(schedule_text_list)


    def get_all_course_sections_with_letter_grades(self) -> None:
        return self.__course_sections_of_student_with_letter_grades

    def get_all_course_sections(self):
        return self.__course_sections_of_student

    def get_all_courses(self) -> List:
        student_courses = {}
        for key,value in self.__course_sections_of_student.items():
            student_courses[key]=value
        return student_courses


    def get_advisor(self) -> Advisor:
        return self.__advisor


    def get_id(self) -> int:
        return self.__id


    def get_name(self) -> str:
        return self.__name


    def get_email(self) -> str:
        return self.__e_mail


    def set_advisor(self, advisor:Advisor) -> None:
        self.__advisor = advisor


    def set_course_section_status(self, course_section, status) -> None:
        self.__course_sections_of_student[course_section]= status
        

    def set_course_section_letter_grade(self, course_section, letter_grade : str) -> None:
            self.__course_sections_of_student_with_letter_grades[course_section]= letter_grade


    def has_pending_course(self) -> bool:
        if 'pending' in self.__course_sections_of_student.values():
            logger.info(f"Student {self.get_id()} has pending courses.")
            return True
        else:
            logger.info(f"Student {self.get_id()} does not have any pending courses.")
            return False

    def get_semester(self)-> int:
        return self.__semester 

    def get_finished_course_credits(self)-> int:
        return sum(course.get_credits() for course in self.get_student_course_sections_with_status("finished"))

    def get_finished_courses(self): # TODO: Create a List for the finished courses 
        return [course_seciton.get_course() for course_seciton in self.get_student_course_sections_with_status("finished") ]


    def print_transcript(self) -> None:
        self.get_controller_instance().menu_operation(f"\n\t\t\tTranscript of {self.__name}\n")
        self.get_controller_instance().menu_operation(f"Year : {math.ceil(self.__semester / 2)}")
        self.get_controller_instance().menu_operation(f"Semester : {self.__semester}")
        self.get_controller_instance().menu_operation(f"\n-----------------------------------------------------------------------------------------------")
        self.get_controller_instance().menu_operation(f"-------------------------------------------------------------------------------------------------\n")
        self.get_controller_instance().menu_operation(self.__transcript.__str__())


    def get_json(self):
        student_data = {
            'AdvisorId': self.get_advisor().get_id(),
            'password': self.get_password(),
            'name': self.get_name(),
            'semester' : self.get_semester(),
            'id': self.get_id(),
            'eMail': self.get_email(),
        }

        course_sections_taken = []
        course_section : CourseSection
        for course_section, status in self.get_all_course_sections().items():
            course_section_data = [
                str(course_section.get_course().get_id()),
                status,
                str(course_section.get_course_section_number()),
                self.get_all_course_sections_with_letter_grades().get(course_section)
            ]
            course_sections_taken.append(course_section_data)
        student_data['CourseSectionsTaken']=course_sections_taken

        return student_data
    